mips/signal.c

```diff
@@ -0,0 +1,86 @@
+#include <signal.h>
+#include <thread.h>
+#include <systm.h>
+#include <stdc.h>
+#include <errno.h>
+
+#define SIG_CTX_MAGIC 0xDACBAEE3
+
+typedef struct sig_ctx {
+  uint32_t magic; /* Integrity control. */
+  exc_frame_t ctx;
+  fpu_ctx_t ctx_fpu;
+  /* TODO: Store signal mask. */
+  /* TODO: Store previous stack data, if the sigaction requested a different
+   * stack. */
+} sig_ctx_t;
+
+/* Delivers a signal to user process. */
+int platform_sendsig(int sig, sigaction_t *sa) {
+  thread_t *td = thread_self();
+  assert(mtx_owned(&td->td_lock));
+
+  /* Prepare signal context. */
+  sig_ctx_t ksc;
+
+  ksc.magic = SIG_CTX_MAGIC;
+  ksc.ctx = td->td_uctx;
+  ksc.ctx_fpu = td->td_uctx_fpu;
+
+  /* Copyout signal context to user stack. */
+  sig_ctx_t *scp = (sig_ctx_t *)td->td_uctx.sp;
+  scp--;
+
+  int error = copyout(&ksc, scp, sizeof(sig_ctx_t));
+  if (error) {
+    /* This thread has a corrupted stack, we can no longer send any signals to
+       it. Forcefully terminate the thread. TODO: Maybe sending a SIGKILL is
+       safe here? */
+    log("Thread %lu is unable to receive a signal, terminating it.",
+        td->td_tid);
+    thread_exit(-1);
+    __builtin_unreachable();
+  }
+
+  /* Prepare user context so that on return to usermode the handler gets
+   * executed. */
+  td->td_uctx.pc = (reg_t)sa->sa_handler;
```

Comment:
I don't think it does hurt not to check if `sa_handler` is a valid address (i.e. aligned and in user-space)... but you should add a comment somewhere it is safe to omit the check.